### ISSUE TITLE
Fail silently

### ISSUE THREAD

**Issue by makdisse** [mayoral/telegrammer#9]
Sorry to open an issue on this, it is more a feature request than an issue I guess.
Do you have any plans to implement a [fail silently](https://github.com/eljojo/telegram_bot/blob/master/lib/telegram_bot/bot.rb#L58) param for the bot.get_updates method?
I'm getting a few timeouts and the app will simply quit with an error.
I wonder if you left it out because you expect each dev to deal with it on their own code?

thanks

**Comment by mayoral** [comment]
What version of the gem are you using?

**Comment by makdisse** [comment]
The gem version is 0.3.1

**Comment by mayoral** [comment]
Can you paste more information about the timeouts, like the backtrace of the error you are receiving?

I'll probably implement something like the fail_silently param, but first I'd like to have a better understanding of what's happening in the get_updates loop to finally receive these timeouts. In fact the migration to httpclient was done to get support for persistent connections.

**Comment by makdisse** [comment]
This is the error message:

```
/usr/lib/ruby/1.9.1/openssl/buffering.rb:53:in `sysread': Interrupt
        from /usr/lib/ruby/1.9.1/openssl/buffering.rb:53:in `fill_rbuff'
        from /usr/lib/ruby/1.9.1/openssl/buffering.rb:200:in `gets'
        from /var/lib/gems/1.9.1/gems/httpclient-2.6.0.1/lib/httpclient/session.rb:356:in `gets'
        from /var/lib/gems/1.9.1/gems/httpclient-2.6.0.1/lib/httpclient/session.rb:876:in `block in parse_header'
        from /usr/lib/ruby/1.9.1/timeout.rb:68:in `timeout'
        from /usr/lib/ruby/1.9.1/timeout.rb:99:in `timeout'
        from /var/lib/gems/1.9.1/gems/httpclient-2.6.0.1/lib/httpclient/session.rb:872:in `parse_header'
        from /var/lib/gems/1.9.1/gems/httpclient-2.6.0.1/lib/httpclient/session.rb:855:in `read_header'
        from /var/lib/gems/1.9.1/gems/httpclient-2.6.0.1/lib/httpclient/session.rb:662:in `get_header'
        from /var/lib/gems/1.9.1/gems/httpclient-2.6.0.1/lib/httpclient.rb:1247:in `do_get_header'
        from /var/lib/gems/1.9.1/gems/httpclient-2.6.0.1/lib/httpclient.rb:1194:in `do_get_block'
        from /var/lib/gems/1.9.1/gems/httpclient-2.6.0.1/lib/httpclient.rb:974:in `block in do_request'
        from /var/lib/gems/1.9.1/gems/httpclient-2.6.0.1/lib/httpclient.rb:1082:in `protect_keep_alive_disconnected'
        from /var/lib/gems/1.9.1/gems/httpclient-2.6.0.1/lib/httpclient.rb:969:in `do_request'
        from /var/lib/gems/1.9.1/gems/httpclient-2.6.0.1/lib/httpclient.rb:822:in `request'
        from /var/lib/gems/1.9.1/gems/httpclient-2.6.0.1/lib/httpclient.rb:725:in `post'
        from /var/lib/gems/1.9.1/gems/telegrammer-0.3.1/lib/telegrammer/bot.rb:400:in `api_request'
        from /var/lib/gems/1.9.1/gems/telegrammer-0.3.1/lib/telegrammer/bot.rb:43:in `block in get_updates'
        from /var/lib/gems/1.9.1/gems/telegrammer-0.3.1/lib/telegrammer/bot.rb:42:in `loop'
        from /var/lib/gems/1.9.1/gems/telegrammer-0.3.1/lib/telegrammer/bot.rb:42:in `get_updates'
        from worker.rb:184:in `<main>'
```

Let me know if is there anything else I can post.
Thanks

**Comment by TheZ3ro** [comment]
After some hours using the 0.4.1 verison I've get this error

It's in the same file as @makdisse but one line after, it's the same sysread function but mine is a HTTPClient::ReceiveTimeoutError and makdisse is an Interrupt (I think it's the same error but different ruby/openssl version)

```
/home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/2.0.0/openssl/buffering.rb:54:in `sysread': execution expired (HTTPClient::ReceiveTimeoutError)
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/2.0.0/openssl/buffering.rb:54:in `fill_rbuff'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/2.0.0/openssl/buffering.rb:201:in `gets'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/httpclient-2.6.0.1/lib/httpclient/session.rb:356:in `gets'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/httpclient-2.6.0.1/lib/httpclient/session.rb:876:in `block in parse_header'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/httpclient-2.6.0.1/lib/httpclient/session.rb:872:in `parse_header'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/httpclient-2.6.0.1/lib/httpclient/session.rb:855:in `read_header'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/httpclient-2.6.0.1/lib/httpclient/session.rb:662:in `get_header'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/httpclient-2.6.0.1/lib/httpclient.rb:1247:in `do_get_header'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/httpclient-2.6.0.1/lib/httpclient.rb:1194:in `do_get_block'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/httpclient-2.6.0.1/lib/httpclient.rb:974:in `block in do_request'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/httpclient-2.6.0.1/lib/httpclient.rb:1082:in `protect_keep_alive_disconnected'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/httpclient-2.6.0.1/lib/httpclient.rb:969:in `do_request'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/httpclient-2.6.0.1/lib/httpclient.rb:822:in `request'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/httpclient-2.6.0.1/lib/httpclient.rb:725:in `post'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/telegrammer-0.4.1/lib/telegrammer/bot.rb:443:in `api_request'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/telegrammer-0.4.1/lib/telegrammer/bot.rb:43:in `block in get_updates'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/telegrammer-0.4.1/lib/telegrammer/bot.rb:42:in `loop'
    from /home/thezero/.rvm/rubies/ruby-2.0.0-p451/lib/ruby/gems/2.0.0/gems/telegrammer-0.4.1/lib/telegrammer/bot.rb:42:in `get_updates'
    from bin/telegrammer-plugin:53:in `<main>'
```

Why HTTPClient keep going timeout if it's using `HTTP/1.1` persistent connections?

Anyway I think this should be a workaround:

``` ruby
begin
    bot.get_updates do |message|
        #Other Code
rescue HTTPClient::ReceiveTimeoutError=>e
    puts "Error: #{e}"
    retry
end
```

Tell me if I'm wrong

**Comment by mayoral** [comment]
@makdisse I'm closing this, since I've released version 0.5.0 with the fail_silently option in get_updates (thanks again to @TheZ3ro for his contribution).

**Comment by TheZ3ro** [comment]
Thanks to you, telegrammer is awesome.

If someone gets a TimeoutError, rescue it and use `retry`
